Implement FileSystemWritableFileStream methods

szewai opened the pull request:
#### d1754516e002176f1e296df7894651446ed8b043
<pre>
Implement FileSystemWritableFileStream methods
<a href="https://bugs.webkit.org/show_bug.cgi?id=284735">https://bugs.webkit.org/show_bug.cgi?id=284735</a>
<a href="https://rdar.apple.com/141528850">rdar://141528850</a>

Reviewed by Youenn Fablet.

Spec: <a href="https://fs.spec.whatwg.org/#dom-filesystemwritablefilestream-write">https://fs.spec.whatwg.org/#dom-filesystemwritablefilestream-write</a>
      <a href="https://fs.spec.whatwg.org/#dom-filesystemwritablefilestream-seek">https://fs.spec.whatwg.org/#dom-filesystemwritablefilestream-seek</a>
      <a href="https://fs.spec.whatwg.org/#dom-filesystemwritablefilestream-truncate">https://fs.spec.whatwg.org/#dom-filesystemwritablefilestream-truncate</a>

* LayoutTests/imported/w3c/web-platform-tests/file-system-access/sandboxed_FileSystemDirectoryHandle-move.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/file-system-access/sandboxed_FileSystemDirectoryHandle-move.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-IndexedDB.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-IndexedDB.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-BroadcastChannel.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-Error.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-MessagePort-frames.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-MessagePort-windows.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-MessagePort-workers.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-frames.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-windows.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-workers.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-remove.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-remove.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemDirectoryHandle-getFileHandle.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemDirectoryHandle-getFileHandle.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemDirectoryHandle-iteration.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemDirectoryHandle-iteration.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemDirectoryHandle-removeEntry.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemDirectoryHandle-removeEntry.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemFileHandle-getFile.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemFileHandle-getFile.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemFileHandle-move.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemFileHandle-move.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemWritableFileStream-write.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemWritableFileStream-write.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemWritableFileStream.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemWritableFileStream.https.any.worker-expected.txt:
* LayoutTests/storage/filesystemaccess/resources/writable-file-stream-seek.js:
(async checkFileContent):
(async runWriterTest):
(async runTest):
(async test):
* LayoutTests/storage/filesystemaccess/resources/writable-file-stream-truncate.js:
(async runWriterTest):
(async runTest):
(async test):
* LayoutTests/storage/filesystemaccess/resources/writable-file-stream-write.js:
(async runTest):
(async test):
* LayoutTests/storage/filesystemaccess/writable-file-stream-seek-expected.txt:
* LayoutTests/storage/filesystemaccess/writable-file-stream-truncate-expected.txt:
* LayoutTests/storage/filesystemaccess/writable-file-stream-write-expected.txt:
* Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStream.cpp:
(WebCore::convertChunk):
(WebCore::FileSystemWritableFileStream::write):
(WebCore::FileSystemWritableFileStream::seek):
(WebCore::FileSystemWritableFileStream::truncate):
* Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStream.h:
(isType):
* Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStream.idl:
* Source/WebCore/Modules/streams/WritableStream.h:
(WebCore::WritableStream::type const):
* Source/WebCore/Modules/streams/WritableStream.idl:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/InternalWritableStream.cpp:
(WebCore::InternalWritableStream::writeChunkForBingings):
* Source/WebCore/bindings/js/InternalWritableStream.h:
* Source/WebCore/bindings/js/JSDOMPromiseDeferred.h:
(WebCore::DOMPromiseDeferredBase::rejectWithCallback):
* Source/WebCore/bindings/js/JSWritableStreamCustom.cpp: Copied from Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStream.idl.
(WebCore::toJS):
(WebCore::toJSNewlyCreated):

Canonical link: <a href="https://commits.webkit.org/289579@main">https://commits.webkit.org/289579@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cf5e87632135c4ed1a22f859d12a6b1bc1776bd7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87274 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6784 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41635 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92135 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38014 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89324 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7060 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/14856 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67444 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25177 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90276 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5408 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78988 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47765 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5180 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33369 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37131 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75662 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34243 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94023 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14436 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10509 "Found 1 new test failure: fast/css/view-transitions-zoom.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76250 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14641 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74844 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75455 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18584 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19796 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18245 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7368 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14455 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/19748 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14200 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17643 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/15981 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/bindings/js/JSFileSystemWritableFileStreamCustom.cpp:
```diff
@@ -0,0 +1,100 @@
+
+static JSC::JSValue writeImpl(FileSystemWritableFileStream& stream, JSC::JSGlobalObject& lexicalGlobalObject, JSC::JSValue data)
+{
+    auto& vm = JSC::getVM(&lexicalGlobalObject);
```
Ref

youennf commented on Source/WebCore/bindings/js/JSFileSystemWritableFileStreamCustom.cpp:
```diff
@@ -0,0 +1,100 @@
+    arguments.append(stream.internalWritableStream().guardedObject());
+    ASSERT(!arguments.hasOverflowed());
+    auto& writerPrivateName = clientData->builtinFunctions().writableStreamInternalsBuiltins().acquireWritableStreamDefaultWriterPrivateName();
+    auto writerOrException = InternalWritableStream::invokeWritableStreamFunction(lexicalGlobalObject, writerPrivateName, arguments);
```
Could we instead move that code to InternalWritableStream instead as a helper routine?

szewai replied:
Moved to `InternalWritableStream`

youennf commented on Source/WebCore/bindings/js/JSFileSystemWritableFileStreamCustom.cpp:
```diff
@@ -0,0 +1,100 @@
+            position = positionConversionResult.returnValue();
+    }
+    FileSystemWritableFileStream::WriteParams params { FileSystemWriteCommandType::Seek, std::nullopt, position, std::nullopt };
+    auto data = toJS<IDLDictionary<FileSystemWritableFileStream::WriteParams>>(lexicalGlobalObject, *JSC::jsCast<JSDOMGlobalObject*>(&lexicalGlobalObject), params);
```
toJS may trigger an exception in workers for termination.
We should probably either check this or make writeImpl check for an existing exception.

szewai replied:
I made data undefined if there is an exception. 
I didn't return early as the [spec](https://fs.spec.whatwg.org/#api-filesystemwritablefilestream-seek) says return the result of "writing a chunk", so I guess we want to invoke `writableStreamDefaultWriterWrite` function (and `FileSystemWritableFileStreamSink::write`) with an invalid parameter.
Let me know if you think this should be done differently.

youennf commented on Source/WebCore/bindings/js/JSFileSystemWritableFileStreamCustom.cpp:
```diff
@@ -0,0 +1,100 @@
+        auto scope = DECLARE_THROW_SCOPE(lexicalGlobalObject.vm());
+        auto sizeConversionResult = convert<IDLUnsignedLong>(lexicalGlobalObject, callFrame.argument(0));
+        if (!sizeConversionResult.hasException(scope))
+            size = sizeConversionResult.returnValue();
```
If there is an exception, we should likely return early

szewai replied:
Same as above: I didn't return early here as it seems should go through "write a chunk" according to spec.

youennf replied:
I am not sure the spec is clear enough here. A method that returns a promise should never throw and always return a promise.
This means that:
- If conversion fails, a rejected promise with `TypeError` should be returned.
- If acquiring the writer throws, a rejected promise with `TypeError`.
See `IDLOperationReturningPromise` for instance.

youennf commented on Source/WebCore/bindings/js/JSFileSystemWritableFileStreamCustom.cpp:
```diff
@@ -0,0 +1,91 @@
+
+JSC::JSValue JSFileSystemWritableFileStream::write(JSC::JSGlobalObject& lexicalGlobalObject, JSC::CallFrame& callFrame)
+{
+    return wrapped().internalWritableStream().writeChunkForBingings(lexicalGlobalObject, callFrame.argument(0));
```
In theory, we should IDL convert callFrame.argument(0) here before locking the writer.
I wonder whether we could do the following:
1. Remove Custom for write, seek and truncate. That way we get IDL conversion for free and a a DOMPromise to resolve/reject
2. Get the write promise as a JSPromise from the write call of InternalWritableStream.
3. Settle the DOMPromise based on the JSPromise via whenSettled.

szewai committed:
Implement FileSystemWritableFileStream methods
https://bugs.webkit.org/show_bug.cgi?id=284735
rdar://141528850

Reviewed by Youenn Fablet.

Spec: https://fs.spec.whatwg.org/#dom-filesystemwritablefilestream-write
      https://fs.spec.whatwg.org/#dom-filesystemwritablefilestream-seek
      https://fs.spec.whatwg.org/#dom-filesystemwritablefilestream-truncate

* LayoutTests/imported/w3c/web-platform-tests/file-system-access/sandboxed_FileSystemDirectoryHandle-move.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/file-system-access/sandboxed_FileSystemDirectoryHandle-move.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-IndexedDB.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-IndexedDB.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-BroadcastChannel.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-Error.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-MessagePort-frames.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-MessagePort-windows.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-MessagePort-workers.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-frames.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-windows.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-workers.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-remove.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-remove.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemDirectoryHandle-getFileHandle.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemDirectoryHandle-getFileHandle.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemDirectoryHandle-iteration.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemDirectoryHandle-iteration.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemDirectoryHandle-removeEntry.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemDirectoryHandle-removeEntry.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemFileHandle-getFile.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemFileHandle-getFile.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemFileHandle-move.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemFileHandle-move.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemWritableFileStream-write.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemWritableFileStream-write.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemWritableFileStream.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemWritableFileStream.https.any.worker-expected.txt:
* LayoutTests/storage/filesystemaccess/resources/writable-file-stream-seek.js:
(async checkFileContent):
(async runWriterTest):
(async runTest):
(async test):
* LayoutTests/storage/filesystemaccess/resources/writable-file-stream-truncate.js:
(async runWriterTest):
(async runTest):
(async test):
* LayoutTests/storage/filesystemaccess/resources/writable-file-stream-write.js:
(async runTest):
(async test):
* LayoutTests/storage/filesystemaccess/writable-file-stream-seek-expected.txt:
* LayoutTests/storage/filesystemaccess/writable-file-stream-truncate-expected.txt:
* LayoutTests/storage/filesystemaccess/writable-file-stream-write-expected.txt:
* Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStream.cpp:
(WebCore::convertChunk):
(WebCore::FileSystemWritableFileStream::write):
(WebCore::FileSystemWritableFileStream::seek):
(WebCore::FileSystemWritableFileStream::truncate):
* Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStream.h:
(isType):
* Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStream.idl:
* Source/WebCore/Modules/streams/WritableStream.h:
(WebCore::WritableStream::type const):
* Source/WebCore/Modules/streams/WritableStream.idl:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/InternalWritableStream.cpp:
(WebCore::InternalWritableStream::writeChunkForBingings):
* Source/WebCore/bindings/js/InternalWritableStream.h:
* Source/WebCore/bindings/js/JSDOMPromiseDeferred.h:
(WebCore::DOMPromiseDeferredBase::rejectWithCallback):
* Source/WebCore/bindings/js/JSWritableStreamCustom.cpp: Copied from Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStream.idl.
(WebCore::toJS):
(WebCore::toJSNewlyCreated):

Canonical link: https://commits.webkit.org/289579@main
